test/jdk/java/lang/StableValue/StableMapTest.java

```diff
@@ -0,0 +1,355 @@
+
+    @Test
+    void factoryInvariants() {
+        assertThrows(NullPointerException.class, () -> StableValue.map(KEYS, null));
+        assertThrows(NullPointerException.class, () -> StableValue.map(null, IDENTITY));
+    }
+
+    @Test
+    void isEmpty() {
+        assertFalse(newMap().isEmpty());
+        assertTrue(newEmptyMap().isEmpty());
+    }
+
+    @Test
+    void size() {
+        assertEquals(KEYS.size(), newMap().size());
+        assertEquals(EMPTY.size(), newEmptyMap().size());
+    }
+
+    @Test
+    void get() {
+        StableTestUtil.CountingFunction<Integer, Integer> cf = new StableTestUtil.CountingFunction<>(IDENTITY);
+        var lazy = StableValue.map(KEYS, cf);
+        int cnt = 1;
+        for (int i : KEYS) {
+            assertEquals(i, lazy.get(i));
+            assertEquals(cnt, cf.cnt());
+            assertEquals(i, lazy.get(i));
+            assertEquals(cnt++, cf.cnt());
+        }
+        assertNull(lazy.get(NOT_PRESENT));
+    }
+
+    @Test
+    void getException() {
+        StableTestUtil.CountingFunction<Integer, Integer> cf = new StableTestUtil.CountingFunction<>(_ -> {
+            throw new UnsupportedOperationException();
+        });
+        var lazy = StableValue.map(KEYS, cf);
+        assertThrows(UnsupportedOperationException.class, () -> lazy.get(KEY));
+        assertEquals(1, cf.cnt());
+        assertThrows(UnsupportedOperationException.class, () -> lazy.get(KEY));
+        assertEquals(2, cf.cnt());
+    }
+
+    @Test
+    void containsKey() {
+        var lazy = newMap();
+        for (int i : KEYS) {
+            assertTrue(lazy.containsKey(i));
+        }
+        assertFalse(lazy.containsKey(NOT_PRESENT));
+    }
+
+    @Test
+    void containsValue() {
+        var lazy = newMap();
+        for (int i : KEYS) {
+            assertTrue(lazy.containsValue(i));
+        }
+        assertFalse(lazy.containsValue(NOT_PRESENT));
+    }
+
+    @Test
+    void forEach() {
+        var lazy = newMap();
+        Set<Map.Entry<Integer, Integer>> expected = KEYS.stream()
+                .map(i -> new AbstractMap.SimpleImmutableEntry<>(i , i))
+                .collect(Collectors.toSet());
+        Set<Map.Entry<Integer, Integer>> actual = new HashSet<>();
+        lazy.forEach((k, v) -> actual.add(new AbstractMap.SimpleImmutableEntry<>(k , v)));
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    void toStringTest() {
+        assertEquals("{}", newEmptyMap().toString());
+        var map = StableValue.map(Set.of(KEY), IDENTITY);
+        assertEquals("{" + KEY + "=.unset}", map.toString());
+        map.get(KEY);
+        assertEquals("{" + KEY + "=" + KEY + "}", map.toString());
+        String actual = newMap().toString();
+        assertTrue(actual.startsWith("{"));
+        for (int key:KEYS) {
+            assertTrue(actual.contains(key + "=.unset"));
+        }
+        assertTrue(actual.endsWith("}"));
```

Comment:
I guess the problem here is that the order of a map is not deterministic?

Reply:
Yes, that is the problem.